docs/spec/lightclient/detection.md

```diff
@@ -0,0 +1,491 @@
+there are two correct full nodes *i* and *j* that are
+  - on different branches, and
+  - primary or secondary,
+
+then the  detector eventually outputs the fork.
+
+**TODO:** We can weaken the above to "the (not-necessarily correct)
+primary provided branch A, and a correct secondary is on branch B". I
+prefer the above as it is slightly less operational.
+
+
+> #### **[LCD-REQ-REP]**
+> If the  detector observes two conflicting headers for height *h*, 
+> it should try to verify both. If both are verified it should report evidence.
+> If the primary reports header *h* and a secondary reports header *h'*,
+>     and if *h'* can be verified based on common root of trust, then
+>     evidence should be generated; 
+> By verifying we mean calling `VerifyToTarget` from the
+> [[verification]] specification.
+
+## Definitions
+
+- A fixed set of full nodes is provided in the configuration upon
+     initialization. Initially this set is partitioned into
+    -  one full node that is the *primary* (singleton set),
+	-  a set *Secondaries* (of fixed size, e.g., 3),
+	-  a set *FullNodes*.
+- A set *FaultyNodes* of nodes that the light client suspects of being faulty; it is initially empty
+
+- *Lightstore* as defined in the [verification specification][verification].
+
+
+
+
+#### **[LCD-INV-NODES]:**
+The detector shall maintain the following invariants:
+   - *FullNodes \intersect Secondaries = {}*
+   - *FullNodes \intersect FaultyNodes = {}*
+   - *Secondaries \intersect FaultyNodes = {}*
+   
+and the following transition invariant
+   - *FullNodes' \union Secondaries' \union FaultyNodes' = FullNodes \union Secondaries \union FaultyNodes*
+
+## Solution
+
+
+
+### Inter Process Communication
+
+
+```go
+func FetchLightBlock(peer PeerID, height Height) LightBlock
+```
+See the [verification specification][verification] for details.
+
+### Auxiliary Functions (Local)
+
+
+```go
+Replace_Secondary(addr Address)
+```
+- Expected precondition
+    - *FullNodes* is nonempty
+- Expected postcondition
+    - addr is moved from *Secondaries* to *FaultyNodes*
+    - an address *a* is moved from *FullNodes* to *Secondaries*
+- Error condition
+    - if precondition is violated
+
+
+
+### From the verifier
+```go
+func VerifyToTarget(primary PeerID, lightStore LightStore,
+                    targetHeight Height) (LightStore, Result)
+```
+See the [verification specification][verification] for details.
+
+
+
+## Solution
+
+Shared data of the light client
+- a pool of full nodes *FullNodes* that have not been contacted before
+- peer set called *Secondaries*
+- primary
+- LightStore
+
+
+
+
+The problem is solved by calling  the function `ForkDetector` with
+a lightstore that contains a light block that has
+just been verified by the verifier. 
+
+
+```go
+func ForkDetector(ls LightStore)  {
+    Forks.Init // initialize a container in which we collect forks
+	for i, s range Secondaries {
+		sh := FetchLightBlock(s,LightStore.LatestVerified().Height)
```

Comment:
Maybe extract `LightStore.LatestVerified()` into a variable.

Reply:
done. I called it `testedLB`. Not sure it is the best name...